dipy/denoise/patch2self.py

```diff
@@ -0,0 +1,273 @@
+    --------
+    cur_x : ndarray
+        Array of patches corresponding to all the volumes except for the
+        held-out volume.
+
+    y : ndarray
+        Array of patches corresponding to the volume that is used a target for
+        denoising.
+    """
+    # Delete the target volume
+    mask = np.zeros(train.shape[0])
+    mask[vol_idx] = 1
+    cur_x = train[mask == 0]
+    cur_x = cur_x.reshape(((train.shape[0]-1)*train.shape[1],
+                           train.shape[2]))
+
+    # Center voxel of the selected block
+    y = train[vol_idx, train.shape[1]//2, :]
+    return cur_x, y
+
+
+def _vol_denoise(train, vol_idx, model, data, alpha):
+    """ Denoise a single 3D volume using a train and test phase.
+
+    Parameters
+    ----------
+    train : ndarray
+        Array of all 3D patches flattened out to be 2D.
+
+    vol_idx: int
+        The volume number that needs to be held out for training.
+
+    model: str
+        Corresponds to the object of the regressor being used for
+        performing the denoising.
+
+    data: ndarray
+        The 4D noisy DWI data to be denoised.
+
+    alpha: float, optional
+        Regularization parameter only for ridge regression model.
+        default: 1.0
+
+    Returns
+    --------
+    model prediction : ndarray
+        Denoised array of all 3D patches flattened out to be 2D corresponding
+        to the held out volume `vol_idx`.
+
+    """
+    # To add a new model, use the following API
+    # We adhere to the following options as they are used for comparisons
+    if model.lower() == 'ols':
+        model = linear_model.LinearRegression(copy_X=False)
+
+    elif model.lower() == 'ridge':
+        model = linear_model.Ridge(copy_X=False, alpha=alpha)
+
+    elif model.lower() == 'lasso':
+        model = linear_model.Lasso(copy_X=False, max_iter=50)
+
+    else:
+        raise ValueError('Model not supported. ',
+                         'Choose from: ols, ridge or lasso',
+                         data.shape)
+
+    cur_x, y = _vol_split(train, vol_idx)
+    model.fit(cur_x.T, y.T)
+
+    return model.predict(cur_x.T).reshape(data.shape[0], data.shape[1],
+                                          data.shape[2])
+
+
+def _extract_3d_patches(arr, patch_radius):
+    """ Extract 3D patches from 4D DWI data.
+
+    Parameters
+    ----------
+    arr : ndarray
+        The 4D noisy DWI data to be denoised.
+
+    patch_radius : int or 1D array
+        The radius of the local patch to be taken around each voxel (in
+        voxels).
+
+    Returns
+    --------
+    all_patches : ndarray
+        All 3D patches flattened out to be 2D corresponding to the each 3D
+        volume of the 4D DWI data.
+
+    """
+    if isinstance(patch_radius, int):
+        patch_radius = np.ones(3, dtype=int) * patch_radius
+    if len(patch_radius) != 3:
+        raise ValueError("patch_radius should have length 3")
+    else:
+        patch_radius = np.asarray(patch_radius, dtype=(int))
```

Comment:
This is not wrong, but it's also not necessary:

```suggestion
        patch_radius = np.asarray(patch_radius, dtype=int)
```